File: solutions/javascript/beer-song/1/beer-song.js
"use strict";
var Beer = (function() {
  function Beer(numberOfBottles) {
    var self = this;
    this.verse = function() {
      return self.bottleText() + ' of beer on the wall, '
             + self.bottleText().toLowerCase() + ' of beer.\n' + self.bottleAction();
    }
    this.bottleText = function() {
      var numberOfBottlesText = numberOfBottles + ' bottles';
      if (numberOfBottles === 1) {
        numberOfBottlesText = numberOfBottles + ' bottle';
      } else if (numberOfBottles === 0) {
        numberOfBottlesText = 'No more bottles';
      }
      return numberOfBottlesText;
    }
    this.bottleAction = function() {
      var nextBeer = self.nextBeer(), nextBottleAction = '';
      if (numberOfBottles > 1) {
        nextBottleAction = 'Take one down and pass it around';
      } else if (numberOfBottles === 1) {
        nextBottleAction = 'Take it down and pass it around';
      } else if(numberOfBottles === 0) {
        nextBottleAction = 'Go to the store and buy some more';
      }
      nextBottleAction += ', ' + nextBeer.bottleText().toLowerCase() + ' of beer on the wall.\n';
      return nextBottleAction;
    }
    this.nextBeer = function() {
      if (numberOfBottles === 0) {
        return new Beer(99);
      } else {
        return new Beer(numberOfBottles - 1);
      }
    }
  }

  Beer.verse = function(verseNumber) {
    var beer = new Beer(verseNumber);
    return beer.verse();
  }

  Beer.sing = function(startVerseNumber, endVerseNumber) {
    var song = [];
    if (typeof endVerseNumber === 'undefined') { endVerseNumber = 0; }
    for (var verseNumber = startVerseNumber; verseNumber >= endVerseNumber; verseNumber--) {
      song.push(Beer.verse(verseNumber));
    }
    return song.join("\n");
  }

  return Beer;
})();

module.exports = Beer;

File: solutions/elixir/beer-song/2/beer-song.exs
defmodule Beer do
  def verse(verse_number), do: first_line(verse_number) <> second_line(verse_number)

  def sing(start_verse_number, finish_verse_number // 0) do
    Enum.map_join(start_verse_number..finish_verse_number, "\n", verse &1) <> "\n"
  end

  defp first_line(verse_number) do
    txt = bottle_text(verse_number)
    "#{txt} of beer on the wall, #{String.downcase(txt)} of beer.\n"
  end

  defp second_line(0) do
    "Go to the store and buy some more, #{next_verse_number(0)} bottles of beer on the wall.\n"
  end
  defp second_line(verse_number) do
    text = next_verse_number(verse_number) |> bottle_text |> String.downcase
    "Take #{take_text(verse_number)} down and pass it around, #{text} of beer on the wall.\n"
  end

  defp next_verse_number(0), do: 99
  defp next_verse_number(verse_number), do: verse_number - 1

  defp bottle_text(0), do: "No more bottles"
  defp bottle_text(1), do: "1 bottle"
  defp bottle_text(bottle_count), do: "#{bottle_count} bottles"

  defp take_text(1), do: "it"
  defp take_text(_), do: "one"
end